Formula grid

gwhitney opened the pull request:
By submitting this PR, I am indicating to the Numberscope maintainers that I have read and understood the [contributing guidelines](https://numberscope.colorado.edu/doc/CONTRIBUTING/) and that this PR follows those guidelines to the best of my knowledge. I have also read the [pull request checklist](https://numberscope.colorado.edu/doc/doc/pull-request-checklist/) and followed the instructions therein.

<hr/>

Introduces the FormulaGrid Visualizer, a formula-based common generalization of Grid and SelfSimilarity.
The following tasks have been completed:

- [x] Implementation of mouseover (prototyped, needs to move near mouse)
- [x] Diagram for hexagonal cells, and guidance for doing triangular arrays.
- [x] Diagram for triangular cells
- [x] Snapshots (we decided on an ulam spiral with the spiral emphasized and the mod-5 arithmetic triangle, and other polishing of documentation ~~(including improving help popups in parameter tab)~~ [decided to leave that for post alpha when we have more features for help facilities, like markdown or context-help "drawer")
- [x] Selection and addition of featured specimens
- [x] Updated tests, including unit tests for triangular and invTriangular
- [x] Investigate phenomenon of specimen tests being erratic on the number of frames they display
- [x] Investigate why aspect ratio does not update when dimensions change
- [x] Make the placeholder text for dimensions and aspect parameters show the values actually being used, together with a phrase like '(from window size)' for dimensions and '(to fill window)' for aspect. 

Here's a classic view of the arithmetic triangle:
http://localhost:5173/?name=OEIS+A007318&viz=FormulaGrid&aspect=r&fillOrder=2&pathFormula=%5Bceil%28c%2F2%29+-+invTriangular%28k-1%29+%2B+2*%28k+-+triangular%28invTriangular%28k-1%29%29+-+1%29%2CinvTriangular%28k-1%29+%2B+1%5D&fillFormula=%7Bhexagon%3A+rainbow%2872a%29%7D&seq=OEIS+A007318

Resolves #527

gwhitney commented:
OK, I added a prototype of mouseover. You can see it in action in this monochrome version of the gcd(x,y) plot (not saying it's the best version of such a plot, by any means):
http://localhost:5173/?name=Formula%3A+n&viz=FormulaGrid&dimensions=200&speed=1024&fillFormula=%7Brectangle%3A+1+-+%28gcd%28x%2Cy%29-1%29%2Fx%2C%0A+mouseover%3A+%5B%0A++++%27gcd%28%27%2C+x%2C+%27%2C+%27%2C+y%2C+%27%29+%3D+%27%2C+gcd%28x%2Cy%29%0A%5D%7D&seq=Formula

However, I definitely don't think it is done. Right now it just overlays the mouseover text at the top right of the canvas, in the default font and color. There's no effort to set it off in a box and no styling. I could definitely use feedback/ideas on what the mouseover text should look like and where it should appear. Should it follow some offset from the mouse like FactorHistogram? Should it always be in the top right corner?  Or should it actually be just above the canvas somewhere so it shows up in the specimen bar? Or some other location option? (Or heavens forfend, should its location be configurable in some way? We should not indulge in configurability always for configurability's sake...)

Personally, I would be happy with it styled in some way (maybe an outlined box with an opaque or semitransparent white or pastel-tinted background, maybe with the same drop shadows as our help boxes) and always appearing in the same place (like the top right where it is now, or any fixed location folks think would be better...).

gwhitney commented:
So I think this PR is now at the stage that it would be best for me to get some feedback before proceeding further, even though it is still a draft.

gwhitney commented:
In particular @katestange if you could try translating some of your specimens from #526 into this FormulaGrid and see how that goes, that would be really great.

katestange commented:
When I preview other sequences, the full set of parameters doesn't seem to be applied in the thumbnails.
![Screenshot from 2025-02-22 17-42-06](https://github.com/user-attachments/assets/279d2816-c665-4d90-8bee-5377d4c457b4)

katestange commented:
With these settings, why is there background showing through?  The aspect ratio of the cells is 'r' for regular, and the "inset" for the squares is '1' to fill the short side.  I put one number in "dimensions" to specify only rows.
![Screenshot from 2025-02-22 17-44-00](https://github.com/user-attachments/assets/63ea8b92-4c6e-4549-96c2-94295f3e022e)

katestange commented:
This particular error may be difficult to isolate, since it presumably has to do with the pile of sequences I have in my sequence chooser, but when I open the sequence chooser, I get `TypeError: args[0].slice is not a function`.  If that doesn't immediately mean something to you, I'll try to track it down to a minimal example.

katestange commented:
This error message seems in error:
![Screenshot from 2025-02-22 17-55-47](https://github.com/user-attachments/assets/236a70c8-44e4-49f0-9eb8-573ebdf107d7)

katestange commented:
This URL is quite laggy:  http://localhost:5173/?name=Formula%3A+n&viz=FormulaGrid&aspect=0.999&backgroundColor=040404FF&speed=100&fillFormula=%7Bcircle%3A+chroma.scale%28%5B%27%23ffff6a%27%2C+%27%23000000%27%5D%29%0A++++.mode%28%27lch%27%29.colors%28600%29%5Bfloor%28601-600*%28gcd%28x%2Cy%29%2Fmax%28x%2Cy%29%29%29%5D%7D&inset=1&seq=Formula

I suspect it's because it has to compute the chroma scale afresh for each dot.

katestange commented:
> I suspect it's because it has to compute the chroma scale afresh for each dot.

Maybe one possibility is to have more inbuilt chroma scales for people to choose from that are pre-computed?

katestange commented:
> I suspect it's because it has to compute the chroma scale afresh for each dot.

This is borne out by a better formula:  http://localhost:5173/?name=Formula%3A+n&viz=FormulaGrid&aspect=0.999&backgroundColor=040404FF&speed=100&fillFormula=%7Bcircle%3A+chroma.mix%28%27%23ffff6a%27%2C+%27%23000000%27%2C+1-gcd%28x%2Cy%29%2Fmax%28x%2Cy%29%2C+%27hsl%27%29%7D&inset=1&seq=Formula

That's a variation on my favourite "gcd" self-similarity.

gwhitney commented:
> With these settings, why is there background showing through?

Because `r` makes the hexagons and triangles regular, which is an aspect ratio of 1/√3. So centering squares in those tall thin rectangles leaves some background visible. If you want to fill the plane with squares, you should use an aspect ratio of 1.

gwhitney commented:
> This error message seems in error: [Cell aspect: 1]
Reproduced. Working on it. Don't push any commits ;-)

gwhitney commented:
> > This error message seems in error: [Cell aspect: 1]
> > Reproduced. Working on it. Don't push any commits ;-)

OK, should be fixed if you pull.

gwhitney commented:
> > I suspect it's because it has to compute the chroma scale afresh for each dot.
> 
> Maybe one possibility is to have more inbuilt chroma scales for people to choose from that are pre-computed?

I observe the same thing, and if I replace the scale with "rainbow" it's super snappy. So I agree, recomputing the scale is expensive. But formula `chroma('#ffff6a').mix( '#000000', 1-gcd(x,y)/max(x,y), 'oklch')` does essentially the same thing, and is shorter and doesn't involve an arbitrary constant 600, and is also super snappy. So I think the moral is don't make very long scales -- when you are doing that, probably you want a continuous mix. Oh and note I changed to mode `oklch` as that is recommended over `lch` in the chroma.js docs -- you can see the difference quite clearly if you switch that string back and forth between `oklch` and `lch` in this visualization.

gwhitney commented:
> This is borne out by a better formula

Ah yes, you and I got to the same place; your comment just hadn't refreshed into my view of this page when I was working on the "laggy" comment

gwhitney commented:
> This particular error may be difficult to isolate, since it presumably has to do with the pile of sequences I have in my sequence chooser, but when I open the sequence chooser, I get `TypeError: args[0].slice is not a function`. If that doesn't immediately mean something to you, I'll try to track it down to a minimal example.

Yeah, I think I would need more than that to debug. I'm not seeing this phenomenon.

gwhitney commented:
> When I preview other sequences, the full set of parameters doesn't seem to be applied in the thumbnails.

I think this is #539 biting you. Let me know if you want me to fold a fix for that into this PR.

gwhitney commented:
Oh, I added a to-do item at the top about featured specimens - @katestange feel free to push a commit with some you like whenever the feeling moves you.

katestange commented:
Would it be ok if I push a few new functions for colours?  There's things I'm trying to do simply/frequently that it would be great to have shortcuts for.  Like "rainbow()".

katestange commented:
It would be great to have a variable for the coordinates of the center of the canvas

katestange commented:
For example, here's Gaussian split prime soup:  (which could use centre coordinates):  http://localhost:5173/?name=Gaussian+Split+Prime+Soup&viz=FormulaGrid&dimensions=100&aspect=0.999&backgroundColor=822F4FFF&speed=100&fillFormula=%7B+circle%3A+chroma%28%27%235db374%27%29.alpha%280.8*isPrime%28%28x-100%29%5E2%2B%28y-50%29%5E2%29%29+%7D&inset=1&seq=Formula

katestange commented:
Another favourite is Modular Multiplication Table: http://localhost:5173/?name=Modular+Multiplication+Table&viz=FormulaGrid&dimensions=100+100&aspect=0.999&backgroundColor=0B207CFF&speed=100&fillFormula=%7B+square%3A+chroma%28%27%233bbf7d%27%29.alpha%28abs%28%28x*y%29+%25+100-50%29%2F50%29+%7D&inset=1&seq=Formula&formula=n%5E2

katestange commented:
Because there's a little question mark next to the little drag handle to grow the formula box, it's hard to get with the mouse.  Maybe the formula box entire lower bar should drag?

katestange commented:
Here's a prime gap spiral: http://localhost:5173/?name=Prime+Gap+Spiral&viz=FormulaGrid&dimensions=100+100&aspect=0.999&fillOrder=1&pathFormula=spiral%28k%29&backgroundColor=FAFBFDFF&speed=100&fillFormula=%7B+square%3A+chroma%28%27%237c0980%27%29.alpha%28%28A%28n%2B1%29-A%28n%29%29%2F50%29+%7D&inset=1&seq=OEIS+A000040

katestange commented:
Mobius Skin Rug: http://localhost:5173/?name=Mobius-Skin+Rug&viz=FormulaGrid&dimensions=200+200&aspect=0.999&fillOrder=1&pathFormula=spiral%28k%29&backgroundColor=FAFBFDFF&speed=100&fillFormula=%7B+square%3A+chroma%28%27%237c0980%27%29.alpha%28%28A%28n%2B1%29-A%28n%29%29%29+%7D&inset=1&seq=OEIS+A008683
It's the differences of consecutive Mobius values on the Ulam spiral...

gwhitney commented:
> Would it be ok if I push a few new functions for colours? There's things I'm trying to do simply/frequently that it would be great to have shortcuts for. Like "rainbow()".

You don't need permission! Just check the chroma.js api doc page to be sure there's not a relatively straightforward way of doing what you want, like `mix` etc.

gwhitney commented:
> It would be great to have a variable for the coordinates of the center of the canvas

`[c/2, r/2]` ??

gwhitney commented:
> Because there's a little question mark next to the little drag handle to grow the formula box, it's hard to get with the mouse. Maybe the formula box entire lower bar should drag?

The browser put the drag handle there, and implements it -- should we move our little question mark help icon?

katestange commented:
Ulam spiral but number of divisors, not just isPrime: http://localhost:5173/?name=Ulam+Divisors&viz=FormulaGrid&dimensions=200+200&aspect=0.999&fillOrder=1&pathFormula=spiral%28k%29&backgroundColor=FAFBFDFF&speed=100&fillFormula=%7B+square%3A+chroma.mix%28%27%230571b0%27%2C+%27darkseagreen%27%0A%2C0.125*%28A%28n%2B1%29-A%28n%29%29%29+%7D&inset=1&seq=OEIS+A000005

katestange commented:
> > This particular error may be difficult to isolate, since it presumably has to do with the pile of sequences I have in my sequence chooser, but when I open the sequence chooser, I get `TypeError: args[0].slice is not a function`. If that doesn't immediately mean something to you, I'll try to track it down to a minimal example.
> 
> Yeah, I think I would need more than that to debug. I'm not seeing this phenomenon.

Ok, from a fresh browser, go to http://localhost:5173/?name=Formula%3A+n&viz=FormulaGrid&dimensions=100+100&aspect=0.999&fillOrder=1&pathFormula=spiral%28k%29&backgroundColor=502020FF&speed=100&fillFormula=%7B+square%3A+chroma.mix%28%27%230571b0%27%2C+%27darkseagreen%27%0A%2C0.0125*%28A%28n%29%29%29+%7D&inset=1&seq=Formula&formula=3%5E%28n%2B2%29-2%5E%28n%2B2%29 and then click on the sequence chooser

It appears that this formula becomes gigantic, and for some reason that causes the popup when it exists in the sequence chooser.

gwhitney commented:
> should we move our little question mark help icon?

In particular, it could be at the top right instead of the bottom right, and I think that would actually make more sense. it seems weird to me that the help icon moves as you drag to resize the input box.

gwhitney commented:
> It appears that this formula becomes gigantic, and for some reason that causes the popup when it exists in the sequence chooser.

Should be fixed now.

katestange commented:
> In particular, it could be at the top right instead of the bottom right, and I think that would actually make more sense. it seems weird to me that the help icon moves as you drag to resize the input box.

That seems reasonable.

gwhitney commented:
> Just pretty

Just so you know you can simplify this:
http://localhost:5173/?name=Stars&viz=FormulaGrid&dimensions=99+99&aspect=1&fillOrder=1&pathFormula=spiral%28k%29&backgroundColor=2D1111FF&fillFormula=chroma.mix%28%27%230571b0%27%2C+%27darkseagreen%27%0A%2C0.00125a%29&inset=1&seq=OEIS+A055684

* `a` is a shorthand for `A(n)`
* If you make the cell aspect 1 and inset 1, then `square` and `rectangle` are the same thing, and if all you have is `rectangle`, you don't even need to wrap it in an object with keys; you can just return the color or array of colors to overlay.

gwhitney commented:
> > In particular, it could be at the top right instead of the bottom right, and I think that would actually make more sense. it seems weird to me that the help icon moves as you drag to resize the input box.
> 
> That seems reasonable.

OK, I will change to that when I get the chance, probably not tonight.

gwhitney commented:
> > > In particular, it could be at the top right instead of the bottom right, and I think that would actually make more sense. it seems weird to me that the help icon moves as you drag to resize the input box.
> > 
> 
> OK, I will change to that when I get the chance, probably not tonight.

OK, I have changed this and taken care of #539 because it kept being a pain. I also changed the formula text area to auto-grow if needed to show the whole formula, when the page is loaded and any time you type anything. Let me know what you think of that change. Also, I'd love to get feedback on the behavior when you add a `mouseover: <WHATEVER TEXT HERE>` property in the fill formula, like this example from above:

http://localhost:5173/?name=Formula%3A+n&viz=FormulaGrid&dimensions=200&speed=1024&fillFormula=%7Brectangle%3A+1+-+%28gcd%28x%2Cy%29-1%29%2Fx%2C%0A+mouseover%3A+%5B%0A++++%27gcd%28%27%2C+x%2C+%27%2C+%27%2C+y%2C+%27%29+%3D+%27%2C+gcd%28x%2Cy%29%0A%5D%7D&seq=Formula

katestange commented:
> Also, I'd love to get feedback on the behavior when you add a `mouseover: <WHATEVER TEXT HERE>` property in the fill formula, like this example from above

Cool!  If you change denominator to y, the text is not very readable... it may need a background rectangle of white.  Also, I would move it slightly left to avoid it hitting up against the right end of the canvas.  Here's the denominator y example to see what I mean: "http://localhost:5173/?name=Formula%3A+n&viz=FormulaGrid&dimensions=200&speed=1024&fillFormula=%7Brectangle%3A+1+-+%28gcd%28x%2Cy%29-1%29%2Fy%2C%0A+mouseover%3A+%5B%0A++++%27gcd%28%27%2C+x%2C+%27%2C+%27%2C+y%2C+%27%29+%3D+%27%2C+gcd%28x%2Cy%29%0A%5D%7D&seq=Formula"

katestange commented:
> > Also, I'd love to get feedback on the behavior when you add a `mouseover: <WHATEVER TEXT HERE>` property in the fill formula, like this example from above

And also, if you'd like to be really fancy, the mouseover text could "hop" out of the way (maybe to the opposite corner) when your mouse goes over there, so you have some way to access the grid pixels covered by the text.

katestange commented:
> K, I have changed this and taken care of #539 because it kept being a pain. I also changed the formula text area to auto-grow if needed to show the whole formula, when the page is loaded and any time you type anything.

This is great!

katestange commented:
I don't know whether you want to go here or not... but the inset could be a formula.

katestange commented on src/visualizers/FormulaGrid.ts:
```diff
@@ -0,0 +1,794 @@
+described above; their values will be supplied by the visualizer. It may
+also use any of the following additional variables:
+
+_k_ -- the 1-based serial number of the cell in the order they are filled in
```
"1-based" again... appears a variety of places

gwhitney replied:
I think I've got them all gone now, except in a code comment.

katestange commented on src/visualizers/FormulaGrid.ts:
```diff
@@ -0,0 +1,794 @@
+Fill formula. If it just returns a color or a value that can be converted to
+a color, the rectangular cell at the current fill position is filled with that
+color. For details on how formulas may create and manipulate colors, see
+the [Chroma documenation](../shared/Chroma.md), but note that a false value
```
documentation

gwhitney replied:
Yup

katestange commented:
Unusual behaviour:  Microsoft edge has totally whacky tooltip behaviour on the parameter input cells.  Depending on how the panel is scrolled up or down by the main scrollbar, some tooltips don't appear, or appear above the parameter pane entirely.

katestange commented:
> Unusual behaviour: Microsoft edge has totally whacky tooltip behaviour on the parameter input cells. Depending on how the panel is scrolled up or down by the main scrollbar, some tooltips don't appear, or appear above the parameter pane entirely. (similar issues in chrome)

Oh I think I know what's happening!  The tooltips are appearing _behind_ the parameter pane instead of on top.

katestange commented on src/visualizers/FormulaGrid.ts:
```diff
@@ -0,0 +1,794 @@
+            + 'keys are rectangle, square, ellipse, circle, hexagon, '
+            + 'triangle, text, or mouseover; the values for the last two '
+            + '"shapes" will be (converted to) strings, not colors. '
+            + formulaDescription,
```
I'm not sure how to improve this pop up description, but the use of this cell is very complicated and a tooltip can't possibly do it justice.  Maybe it should just say something like "A formula giving the color of the cell, or a more complicated formula describing shapes, text and mouseover information for the cell."  Because they need to visit the docs, maybe the only goal here is to describe the function, not the format.  Perhaps this comment could even apply to some of the other tooltips also.

gwhitney replied:
Deferring this for in-person conversation.

katestange commented on src/shared/math.ts:
```diff
@@ -154,6 +159,11 @@ math.typed.addType({
+    triangular: (n: number) => (n * (n + 1)) / 2,
+    invTriangular: (t: number) => Math.floor((Math.sqrt(1 + 8 * t) - 1) / 2),
+}
+
```
should these have unit tests?

gwhitney replied:
Yes they should, will add to todo list.

gwhitney commented:
> > > Also, I'd love to get feedback on the behavior when you add a `mouseover: <WHATEVER TEXT HERE>` property in the fill formula, like this example from above
> 
> And also, if you'd like to be really fancy, the mouseover text could "hop" out of the way (maybe to the opposite corner) when your mouse goes over there, so you have some way to access the grid pixels covered by the text.

Is this what you meant (if you pull)? Also, it is opaque, bordered, and shadowed now, so I think it stands out pretty well wherever it is.

gwhitney commented:
> I don't know whether you want to go here or not... but the inset could be a formula.

Your wish is my code. Here's a version of your favorite gcd grid, but with the brighter circles smaller in diameter so that the overall intensity is relatively constant, just more concentrated at the high-gcd places:
http://localhost:5173/?name=Formula%3A+n&viz=FormulaGrid&backgroundColor=040404FF&speed=100&fillFormula=%7Bcircle%3A+chroma.mix%28yellow%2C+black%2C+1-gcd%28x%2Cy%29%2Fmax%28x%2Cy%29%2C+%27hsl%27%29%7D&inset=%281-gcd%28x%2Cy%29%2Fmax%28x%2Cy%29%29+%2B+0.2&seq=Formula

(not saying this is any better than the one with constant-size dots; mostly just a demo of the inset formula.

gwhitney commented:
> > Unusual behaviour: Microsoft edge has totally whacky tooltip behaviour on the parameter input cells. Depending on how the panel is scrolled up or down by the main scrollbar, some tooltips don't appear, or appear above the parameter pane entirely. (similar issues in chrome)
> 
> Oh I think I know what's happening! The tooltips are appearing _behind_ the parameter pane instead of on top.

OK, edge should be fixed now. Thanks for checking!

gwhitney commented on src/visualizers/FormulaGrid.ts:
```diff
@@ -0,0 +1,794 @@
+then two cells on its left and right in the second row, and so on. In addition,
+you can customize the fill order with a formula giving the coordinates of the
+_k_th cell to be visited starting from _k_ = 1. Note that there is no
+requirement that a given cell in the grid is visited exactly once as a result
```
Yeah, the only way I could find to fix this is to switch from `_k_th` to `_k_-th` which I don't like as well but I think is pretty clear.

gwhitney commented on src/visualizers/FormulaGrid.ts:
```diff
@@ -0,0 +1,794 @@
+    /** md
+- Inset: The side length of the 'square' shape and/or diameter of the 'circle'
+  shape, as a multiple of the shorter dimension of a cell. Must be positive,
+  but allowed to be greater than one, in which case the shapes may overlap.
```
The triangles and hexagons are fixed sizes that singly and doubly cover the plane, respectively.

katestange replied:
Oh, ok.  Is it worth mentioning it is expected to have no effect on those?

gwhitney replied:
OK, did that in the latest commit.

gwhitney commented on src/visualizers/FormulaGrid.ts:
```diff
@@ -0,0 +1,794 @@
+  and `y` giving the coordinates. The formula may use any of the variables
+  listed in the description above, where the inputs _x_, _y_, and _s_ are
+  calculated as if the fill order were the default 'By_Rows'. This latter
+  convention is useful if you want to tweak the default fill order.
```
OK, will fill in an example of each format (for doing vertical instead of horizontal stripes).

gwhitney commented on src/visualizers/FormulaGrid.ts:
```diff
@@ -0,0 +1,794 @@
+
+On the other hand, the fill formula may instead return a plain object with keys
+drawn from the list of supported shapes:
+
```
ok, putting in three examples: one with just a color, one with rectangle and square together, and one with hexagon and mouseover.

gwhitney commented on src/visualizers/FormulaGrid.ts:
```diff
@@ -0,0 +1,794 @@
+  not return an object.
+
+square
+: Draws a square with side a configurable fraction of the shorter dimension
```
ok, will clarify.

gwhitney commented on src/visualizers/FormulaGrid.ts:
```diff
@@ -0,0 +1,794 @@
+  if you take the odd cells horizontally in one row and the even cells in
+  the next, those hexagons tile the plane. Note the Triangle pre-set fill
+  order only visits such cells. When the aspect ratio of a cell is 1/√3, the
+  hexagons are regular. (Needs diagram.)
```
Putting on todo list since I can't get to the diagram tonight.

gwhitney replied:
OK, how does the diagram look? I have also added to the to-do list to add a diagram for the triangles.

gwhitney commented on src/visualizers/FormulaGrid.ts:
```diff
@@ -0,0 +1,794 @@
+
+mouseover
+: Another special shape key. The value of the associated formula will be
+  converted into a string and shown in a popup with the mouse pointer hovers
```
Righto

gwhitney commented:
OK, got most of the review comments, except some I explicitly deferred either because I can't get to them tonight or because I think a voice/video conversation would be helpful.

katestange commented:
> (not saying this is any better than the one with constant-size dots; mostly just a demo of the inset formula.

Very cool!  Love it.

katestange commented:
> Is this what you meant (if you pull)? Also, it is opaque, bordered, and shadowed now, so I think it stands out pretty well wherever it is.

Ok, it's totally legible now!  And matches the decor.  But now that I see it jumping around, I wonder why not just put it near the mouse cursor (where your eyes are looking anyway)?  More like a typical mouseover popup.

katestange commented:
> OK, edge should be fixed now. Thanks for checking!

Agree, looking good.

katestange commented:
> Because `r` makes the hexagons and triangles regular, which is an aspect ratio of 1/√3. So centering squares in those tall thin rectangles leaves some background visible. If you want to fill the plane with squares, you should use an aspect ratio of 1.

BTW, this was because I didn't read the instructions carefully.  So all is well.

gwhitney commented:
> > Is this what you meant (if you pull)? Also, it is opaque, bordered, and shadowed now, so I think it stands out pretty well wherever it is.
> 
> Ok, it's totally legible now! And matches the decor. But now that I see it jumping around, I wonder why not just put it near the mouse cursor (where your eyes are looking anyway)? More like a typical mouseover popup.

OK, I added a to-do to put the popup by the mouse cursor. My plan is to put it below and to the right of the cursor, unless that would move it off of the canvas, in which case it will rotate around the pointer as you approach the bottom or the right edge of the canvas. If that doesn't sound good, let me know.

katestange commented:
Is there something suspicious happening in the last few rows here?  
http://localhost:5173/?name=Pascal%27s+Triangle&viz=FormulaGrid&aspect=r&fillOrder=2&pathFormula=%5Bceil%28c%2F2%29+-+invTriangular%28k-1%29+%2B+2*%28k+-+triangular%28invTriangular%28k-1%29%29+-+1%29%2CinvTriangular%28k-1%29+%2B+1%5D&backgroundColor=21218FFF&fillFormula=%7Bhexagon%3A+rainbow%2872a-80%29.desaturate%281.2%29.brighten%280.8%29%7D&seq=OEIS+A007318

![image](https://github.com/user-attachments/assets/8319668b-1578-43ab-b8e5-a795ffbc94a9)

gwhitney commented:
> Is there something suspicious happening in the last few rows here? 

Presumably number MAX_SAFE_INTEGER overflow messing with the mods.

katestange commented:
The featured specimens e2e tests are being flaky.  In particular, some of them seem to sometimes show one more frame than other times.  I thought that was controlled for...

gwhitney commented:
The specimens look great. Nice to have an example of a point plot. I am putting onto the to-do list investigating non-reproducibility of specimen tests -- I agree, they are supposed to do an exact number of frames.

gwhitney commented:
Allright, the mouseover popup goes with the mouse pointer. Let me know what you think about the behavior at the bottom and right edges and in the bottom-right corner.

katestange commented:
> Allright, the mouseover popup goes with the mouse pointer. Let me know what you think about the behavior at the bottom and right edges and in the bottom-right corner.

I get a typscript error:
```

> frontscope@0.3.0 dev
> make -f etc/Makefile vueserve

npx vue-tsc --noEmit -p tsconfig.vitest.json --composite false
src/visualizers/FormulaGrid.ts:832:23 - error TS18047: 'where' is possibly 'null'.

832             const r = where.getBoundingClientRect()
                          ~~~~~


Found 1 error in src/visualizers/FormulaGrid.ts:832

make: *** [etc/Makefile:39: etc/typechecked_at] Error 2
```

katestange commented:
The mouseover is working wonderfully now!  It seems to go all the way to the edges/corners in all browsers and behave well.

gwhitney commented:
Great! I think all that remains until we have a chance to discuss what the description popups for the parameters should be like is the erraticness you observed in end-to-end testing. I will see if I can reproduce and work on it if so, or let you know if I can't reproduce it so that maybe you can find a more specific way for me to evoke it.

gwhitney commented:
Indeed, in my very first e2e run, one of the formulagrid filled more pixels than expected and one filled less. Everything else was fine. Will investigate.

gwhitney commented:
@katestange, you have the alpha formula in the Modular Multiplication Table featured specimen to be `abs((x*y) % 200-100)/100`, but this seems to presuppose that the number of rows and columns is 200. But a nice thing to do is to experiment with different moduli by changing the number of rows and columns. So would it be better for the formula to be `abs(2((x*y) % c) - c)/c` ?

gwhitney commented:
OK, I believe I tracked down an async race condition between onBeforeMount and onMounted in the Scope.vue view, and fixed it. The tests seem much more solid to me now. Give them a whirl on your end. If all seems well, then we just need to discuss the popups with the description of the formula parameters, and then I need to tweak the documentation just a little more, and update the CI snapshots, and we shoulld be ready for final review.

gwhitney commented:
OK, I had found another bug in that when you changed the dimensions, the canvas wouldn't change shape even if the change resulted in a different requestedAspectRatio. That's fixed now, so we remain in exactly the state of my immediately preceding comment.

katestange commented:
Another cool Ulam spiral: http://localhost:5050/?name=OEIS+A010051&viz=FormulaGrid&dimensions=100+100&aspect=1&fillOrder=1&pathFormula=spiral%28k%29&backgroundColor=000000FF&speed=100&fillFormula=%7B+rectangle%3A+chroma%28%27%233bbf7d%27%29.alpha%28abs%28%28%28isPrime%28k%2B1%29*8%2BisPrime%28k%2B2%29*7%2BisPrime%28k%2B3%29*6%2BisPrime%28k%2B4%29*5%2BisPrime%28k%2B5%29*4%2BisPrime%28k%2B6%29*3%2BisPrime%28k%2B7%29*2%2BisPrime%28k%2B8%29*1%29%29%29%2F20%29+%2C+circle%3A+chroma%28%27yellow%27%29.alpha%28a%29%2C+mouseover%3A+%5B%0A++++%27d%28%27%2C+n%2C+%27%29+%3D+%27%2C+a%0A%5D%7D&inset=0.8&seq=OEIS+A010051

katestange commented:
> Another cool Ulam spiral: http://localhost:5050/?name=OEIS+A010051&viz=FormulaGrid&dimensions=100+100&aspect=1&fillOrder=1&pathFormula=spiral%28k%29&backgroundColor=000000FF&speed=100&fillFormula=%7B+rectangle%3A+chroma%28%27%233bbf7d%27%29.alpha%28abs%28%28%28isPrime%28k%2B1%29*8%2BisPrime%28k%2B2%29*7%2BisPrime%28k%2B3%29*6%2BisPrime%28k%2B4%29*5%2BisPrime%28k%2B5%29*4%2BisPrime%28k%2B6%29*3%2BisPrime%28k%2B7%29*2%2BisPrime%28k%2B8%29*1%29%29%29%2F20%29+%2C+circle%3A+chroma%28%27yellow%27%29.alpha%28a%29%2C+mouseover%3A+%5B%0A++++k%0A%5D%7D&inset=0.8&seq=OEIS+A010051

This is the one Aaron and Glen collaborated one:  http://localhost:5173/?name=OEIS+A010051&viz=FormulaGrid&dimensions=70+70&aspect=1&fillOrder=1&pathFormula=spiral%28k%29&backgroundColor=000000FF&fillFormula=%7Brectangle%3A+chroma.lch%2832+%2B+7*sin%285%2F9*pi*sqrt%28k%29%29%2C+31%2C+100*sqrt%28k%29%29%2C+circle%3A+chroma.lch%2875%2C+81%2C+72%29.alpha%28a%29%7D&inset=0.8&seq=OEIS+A010051

I think it's fantastic, we should add it as a gallery specimen!

katestange commented:
> @katestange, you have the alpha formula in the Modular Multiplication Table featured specimen to be `abs((x*y) % 200-100)/100`, but this seems to presuppose that the number of rows and columns is 200. But a nice thing to do is to experiment with different moduli by changing the number of rows and columns. So would it be better for the formula to be `abs(2((x*y) % c) - c)/c` ?

Definitely!

gwhitney commented:
> I think it's fantastic, we should add it as a gallery specimen!

OK, I will definitely use it for the snapshot. As far as featured specimens go, I think we shouldn't have both the Ulam Divisors and this one -- I feel like they are cognitively too close. I think we may at some point want to have a "library" of specimens that are not all "Featured" and then I'd be happy to include this (as well as maybe a very plain, very zoomed-out Ulam spiral). For now I will just put it in the interesting specimens discussion.

gwhitney commented:
OK, please check out the new placeholders for dimensions and aspect. It was a bit tricky because the placeholder needs to stay even when you have entered the number of rows but not the number of columns, so that you can see how many columns the visualizer ended up choosing. But I think I have it working OK now -- happy for any feedback. As far as I know, once we are settled on this I can update the CI snapshots and mark this as ready for review. Of course if you see anything else that needs improving, just let me know.

katestange commented:
> OK, please check out the new placeholders for dimensions and aspect. It was a bit tricky

Wow, this is great!  It's really nice and intuitive.

gwhitney commented:
OK, I am not at the moment aware of anything that needs to be done with respect to this PR and all tests are passing, so I have marked it ready for review. Of course, if anything comes up in the final review, I am happy to make further changes/fixes.

katestange commented:
Small detail:  "visualizer" isn't capitalized in the documentation page title (in contrast to other Visualizers).  That's the only nitpick I have left!

gwhitney commented:
OK, typo should be fixed.
